pkg/operator/reconciler_integrations.go

```diff
@@ -0,0 +1,118 @@
+package operator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	gragent "github.com/grafana/agent/pkg/operator/apis/monitoring/v1alpha1"
+	"github.com/grafana/agent/pkg/operator/clientutil"
+	"github.com/grafana/agent/pkg/operator/config"
+	apps_v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (r *reconciler) newIntegrationsDeploymentSecret(
+	ctx context.Context,
+	l log.Logger,
+	d gragent.Deployment,
+) error {
+
+	// The Deployment for integrations only has integrations where AllNodes is
+	// false.
+	d = deploymentIntegrationSubset(d, false)
+
+	name := fmt.Sprintf("%s-integrations-deploy-config", d.Agent.Name)
+	return r.createTelemetryConfigurationSecret(ctx, l, name, d, config.IntegrationsType)
+}
+
+func (r *reconciler) newIntegrationsDaemonSetSecret(
+	ctx context.Context,
+	l log.Logger,
+	d gragent.Deployment,
+) error {
+
+	// The DaemonSet for integrations only has integrations where AllNodes is
+	// true.
+	d = deploymentIntegrationSubset(d, true)
+
+	name := fmt.Sprintf("%s-integrations-ds-config", d.Agent.Name)
+	return r.createTelemetryConfigurationSecret(ctx, l, name, d, config.IntegrationsType)
+}
+
+func deploymentIntegrationSubset(d gragent.Deployment, allNodes bool) gragent.Deployment {
+	res := *d.DeepCopy()
+
+	filteredIntegrations := make([]gragent.IntegrationsDeployment, 0, len(d.Integrations))
+	for _, i := range d.Integrations {
+		if i.Instance.Spec.Type.AllNodes == allNodes {
+			filteredIntegrations = append(filteredIntegrations, i)
+		}
+	}
+
+	res.Integrations = filteredIntegrations
+	return res
+}
+
+func (r *reconciler) newIntegrationsDeployment(
+	ctx context.Context,
+	l log.Logger,
+	d gragent.Deployment,
+) error {
+
+	// The Deployment for integrations only has integrations where AllNodes is
+	// false.
+	d = deploymentIntegrationSubset(d, false)
+
+	name := fmt.Sprintf("%s-integrations-deploy", d.Agent.Name)
+	deploy, err := newIntegrationsDeployment(r.config, name, d)
+	if err != nil {
+		return fmt.Errorf("failed to generate integrations Deployment: %w", err)
+	}
+	key := types.NamespacedName{Namespace: deploy.Namespace, Name: deploy.Name}
+
+	if len(d.Integrations) == 0 {
+		level.Info(l).Log("msg", "deleting integrations Deployment", "deploy", key)
+		var deploy apps_v1.Deployment
+		return deleteManagedResource(ctx, r.Client, key, &deploy)
```

Comment:
Can you clarify why the calls to deleteManagedResource are always necessary when this condition is true? Is it possible that `len(d.Integrations)` was never > 0 when this block of code is called?